### PR TITLE
New/add spider chi_ssa_35 and it's test case

### PR DESCRIPTION
**Pull request by sosolidkk**
## Summary

**Issue:** #568 

## Checklist

All checks are run in [GitHub Actions](https://github.com/features/actions). You'll be able to see the results of the checks at the bottom of the pull request page after it's been opened, and you can click on any of the specific checks listed to see the output of each step and debug failures.

- [X] Tests are implemented
- [X] All tests are passing
- [X] Style checks run (see [documentation](https://cityscrapers.org/docs/development/) for more details)
- [X] Style checks are passing
- [ ] Code comments from template removed

## Questions

I am having some doubts about this spider, because it has all the meetings time, date and place displayed on the website itself, but the meeting details for the current day that will happen are inside a `.pdf` document. So what i did was to put the `.pdf` document content displayed into the description field in the spider. Anyway, i don't know if what i did was the correct approach or if the right way would be to iterate over the `.pdf` documents and parse the data inside them as meetings.

### CODE REVIEW

File: city_scrapers/spiders/chi_ssa_35.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    def parse(self, response):
+        """
+        `parse` should always `yield` Meeting items.
+
+        Change the `_parse_title`, `_parse_start`, etc methods to fit your scraping
+        needs.
+        """
+        content_div = response.css("div.content_block.content.background_white")
+        _ = content_div.css("h4::text").getall()
```
Comment (luchiago) [inline city_scrapers/spiders/chi_ssa_35.py:29]: Why are you ignoring this variable?

Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:29]: I was using it for the commission titles in case i needed it later, but since they were all the same i just ignore it.

File: city_scrapers/spiders/chi_ssa_35.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            yield meeting
+
+    def _parse_title(self, item):
+        """Parse or generate meeting title."""
+        return "Commission"
+
+    def _parse_description(self, item):
+        """Parse or generate meeting description."""
+        item_url = item[1].css("::attr(href)").get()
+        response = requests.get(item_url)
```
Comment (pjsier) [inline city_scrapers/spiders/chi_ssa_35.py:60]: There isn't much meaningful information in the description, so we can drop this. We would want to include PDF parsing within scrapy so we can manage rate-limiting and handling responses in a unified way instead of using `requests` here as well

File: city_scrapers/spiders/chi_ssa_35.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        elif len(date_item) == 3:
+            date_item.append("09:00 am")
+            date_item.append(f"{datetime.today().year}")
+
+        return " ".join(date_item)
+
+    def _parse_start(self, item):
+        """Parse start datetime as a naive datetime object."""
+        date_item = self._clean_date_item(item[0])
+        date_obj = datetime.strptime(date_item, "%A, %B %d %H:%M %p %Y")
```
Comment (pjsier) [inline city_scrapers/spiders/chi_ssa_35.py:90]: We've had some issues in the past with agencies putting the wrong day of the week, but the correct date, so we should ignore the weekday. Also, if we're using `%p` we'll need to use `%I` for the hour format

File: city_scrapers/spiders/chi_ssa_35.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        return ""
+
+    def _parse_all_day(self, item):
+        """Parse or generate all-day status. Defaults to False."""
+        return False
+
+    def _parse_location(self, item):
+        """Parse or generate location."""
+        return {
+            "address": "Lincoln Park Chamber of Commerce, 2468 N. Lincoln, Chicago",
```
Comment (pjsier) [inline city_scrapers/spiders/chi_ssa_35.py:109]: It looks like this has varied a lot according to their site. We might want to just include "Confirm with agency" in the `name` and leave `address` blank

File: city_scrapers/spiders/chi_ssa_35.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            "name": "Lincoln Park Chamber of Commerce",
+        }
+
+    def _parse_links(self, item):
+        """Parse or generate links."""
+        href = item[1].css("::attr(href)").get()
+        title = item[1].css("::text").get()
```
Comment (pjsier) [inline city_scrapers/spiders/chi_ssa_35.py:116]: The date isn't very relevant if we're already associating it with a meeting on a given date, so we should label it `Minutes` or `Agenda` based off of the section or URL text

File: city_scrapers/spiders/chi_ssa_35.py (hunk 1 above)
Comment (pjsier) [reply city_scrapers/spiders/chi_ssa_35.py:29]: This can be removed if it isn't being used

**Review by pjsier — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some comments here

File: city_scrapers/spiders/chi_ssa_35.py (hunk 1 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:29]: Ok, i'll remove

File: city_scrapers/spiders/chi_ssa_35.py (hunk 3 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:90]: Ok, i'll make the changes

File: city_scrapers/spiders/chi_ssa_35.py (hunk 4 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:109]: 👍🏻

File: city_scrapers/spiders/chi_ssa_35.py (hunk 5 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:116]: Could you elaborate a little more? I didn't understand @pjsier

Comment (pjsier) [reply city_scrapers/spiders/chi_ssa_35.py:116]: Sure, if we're looking at the document in the context of a meeting we already know the meeting date, so it isn't adding information as the document title. It's more helpful to see whether we're looking at an Agenda or Minutes, so ideally we could check which section of links we're in and assign the title that way, otherwise it looks like "Agenda" or "Minutes" are often included in the URL string so we could parse them that way

Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:116]: Oh, i see. I'll check that out and make all the changes. Thank you for the explanation

**Comment by sosolidkk** [comment]
Hello @pjsier , I was updating some stopped code and I made the corrections suggested by you. I also updated the code to make the year of each item correct, since it was fixed with a `datetime.today().year`. The only problem I still have is your change suggestion to `Minutes` and `Agenda` on title. I can't think of a way to do this dynamically, since on the page all I have is a `<h4>` which is followed by several `<p>` tags that contain the links inside. I kind of have to count and make it a more hard coded process. Do you have any better suggestions?

**Comment by pjsier** [comment]
@sosolidkk thanks for the changes! I mentioned in the comment, but the `href` attribute usually contains "Agenda" or "Minutes" which is one way, and you could also loop through a selector that iterates through the immediate children of `.content` and updates the document name any time it runs into an `h4`

**Comment by sosolidkk** [comment]
Hey @pjsier , sorry for the delay. I've updated this PR with the changes that you request. Now i'm iterating over all the inner elements of the body and separating the items in groups based on their `<h4>` title value, that can be Agenda, Schedule or Minutes.

File: city_scrapers/spiders/chi_ssa_35.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+            element_content["description"] = _description
+            element_content["type"] = _type
+            element_content["year"] = _year
+            content.append(element_content)
+
+        self._add_year_to_date_item(content)
+        self._parse_date_to_datetime(content)
+
+        for item in content:
```
Comment (pjsier) [inline city_scrapers/spiders/chi_ssa_35.py:61]: Is there a reason for the two separate looks here? It seems like we could consolidate them and have the individual parse functions handle getting details from a parent element rather than as separate steps of a loop

File: city_scrapers/spiders/chi_ssa_35.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                links=self._parse_links(item),
+                source=self._parse_source(response),
+            )
+
+            meeting["status"] = self._get_status(meeting)
+            meeting["id"] = self._get_id(meeting)
+
+            yield meeting
+
+    def _add_year_to_date_item(self, content):
```
Comment (pjsier) [inline city_scrapers/spiders/chi_ssa_35.py:83]: It's a little hard to follow this when it's modifying an object in-place and not returning a value, could this be made more explicit in a function that returns a value instead of modifying one?

**Review by pjsier — CHANGES_REQUESTED** [review]
Sorry for the delay, and thanks for the updates! I had a few general questions I wanted to address first, let me know if I can clarify anything

File: city_scrapers/spiders/chi_ssa_35.py (hunk 7 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:83]: Sure, I could do that.

File: city_scrapers/spiders/chi_ssa_35.py (hunk 6 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:61]: I actually don't have a reason for using two for loops. For me it was more simpler to save all data into a list and the just run the data that is into the list, but i can change that too. Anyways, the suggestion that you made about search for the elements and iterating over them was just like the way i did? Or did you have something else in mind?

File: city_scrapers/spiders/chi_ssa_35.py (hunk 7 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:82]: Done!

File: city_scrapers/spiders/chi_ssa_35.py (hunk 6 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py]: Done!

File: city_scrapers/spiders/chi_ssa_35.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                meeting_content["date"] = element.css("::text").get()
+                if "href" in element.get():
+                    meeting_content["url"] = element.css("a::attr(href)").get()
+
+            if "<p><a" in element.get():
+                meeting_content["date"] = element.css("::text").get()
+                meeting_content["url"] = element.css("a::attr(href)").get()
+
+            meeting_content["description"] = _description
+            meeting_content["type"] = _type
```
Comment (pjsier) [inline city_scrapers/spiders/chi_ssa_35.py:53]: It looks like this isn't currently being used?

File: city_scrapers/spiders/chi_ssa_35.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            meeting_content["year"] = _year
+
+            if "date" not in meeting_content:
```
Comment (pjsier) [inline city_scrapers/spiders/chi_ssa_35.py:56]: There may be some meetings that are both listed for agendas and minutes, and we'll need to combine those so that this doesn't return duplicates.

**Review by pjsier — CHANGES_REQUESTED** [review]
Thanks for these updates, it's coming along! I left a few comments here

File: city_scrapers/spiders/chi_ssa_35.py (hunk 8 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:53]: It isn't. I was looking for somewhere to put this info.

File: city_scrapers/spiders/chi_ssa_35.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        """Parse start datetime as a naive datetime object."""
+        return item["date_obj"]
+
+    def _parse_end(self, item):
+        """Parse end datetime as a naive datetime object. Added by pipeline if None"""
+        return item["date_obj"] + timedelta(minutes=90)
+
+    def _parse_time_notes(self, item):
+        """Parse any additional notes on the timing of the meeting"""
+        return ""
```
[the remark this answers is not in the document]
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:125]: Are you talking about the `contact us` section?

```
Contact Us
Lincoln Park Chamber of Commerce
2468 N. Lincoln
Chicago, IL 60614
773 880 5200
```

File: city_scrapers/spiders/chi_ssa_35.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    def _parse_location(self, item):
+        """Parse or generate location."""
+        return {
+            "address": "",
+            "name": "Confirm with agency",
+        }
+
+    def _parse_links(self, item):
+        """Parse or generate links."""
+        return [{"href": item.get("url"), "title": item["date"]}]
```
[the remark this answers is not in the document]
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:140]: Can i just replace the title key with the info needed? Agenda or minutes

File: city_scrapers/spiders/chi_ssa_35.py (hunk 9 above)
Comment (sosolidkk) [reply city_scrapers/spiders/chi_ssa_35.py:56]: Ok

**Review by simran-2501 — APPROVED** [review]
